rfcs/20210129-optimize-pd-creation/README.md

```diff
@@ -0,0 +1,262 @@
+    op_desc_t *op_desc_;
+    primitive_attr_t *attr_;
+    std::type_index impl_id_;
+    std::vector<memory_desc_t> bwd_mds;
+
+    // Obtained from `op_desc_t`
+    primitive_kind_t primitive_kind_;
+
+    // Obtained from `engine_t`
+    engine_kind_t engine_kind_;
+    runtime_kind_t runtime_kind_;
+    device_id_t device_id_;
+
+    // Obtained from ctor
+    int impl_nthr_;
+
+    // ... //
+```
+
+When creating a key to find a primitive in the cache there is a primitive descriptor
+which is used to obtain some of the required data. However, in case of primitive descriptor
+creation for all primitives except reorder, concat and sum user creates `op_desc_t`,
+`primitive_attr_t`, `engine_t` and optionally a hint primitive descriptor. Therefore
+`impl_id_`, `bwd_mds` and `op_desc_t` for the aforementioned primitives cannot
+be obtained at that point.
+
+#### Implementation ID
+
+The rationale behind having an implementation ID is to be able to differentiate
+the same primitives which have different underlying implementations. Such primitives
+can be created when user uses primitive descriptor iterator to manually iterate
+through the list of implementations.
+
+The problem here is that the primitive descriptor is only going to be created and
+currently there is no way to get implementation ID without creating primitive
+descriptor.
+
+Currently, `std::type_index` of a primitive descriptor of a particular implementation
+is used as the implementation ID. In general, this approach has only one minor issue,
+it doesn't make it possible to build oneDNN without RTTI support, though it's
+unclear whether such ability is needed. However, to be able to use the existing
+key and hence the existing cache for primitives to speed up primitive descriptor
+creation the implementation ID must be reworked.
+
+The idea is to replace the existing implementation ID with a new one which is
+defined as follows:
+
+Implementation ID is a relative index of a valid implementation for the given
+operation descriptor, attributes and a forward primitive descriptor hint.
+
+For example there are operation descriptor and attributes for which some of
+the implementations in the implementation list are valid. Then the implementation
+IDs will be the following:
+
+| Implementation ID | Implementation index in the list | Valid |
+| ----------------- | -------------------------------- | ----- |
+|       -           |                0                 |   -   |
+|       0           |                1                 |   +   |
+|       -           |                2                 |   -   |
+|       1           |                3                 |   +   |
+|       -           |                4                 |   -   |
+|       2           |                5                 |   +   |
+
+In other words, the implementation IDs are the indexes in a hypothetical list
+of valid implementations for the given operation descriptors and attributes.
```

Comment:
I am not a huge fan of this approach since it ties the primitive cache to our dispatching mechanism.
If I understand correctly, if we want to change our dispatching mechanism, this would require to rework how the primitive (descriptor) cache works as well under the current proposal.

Reply:
It's not something that is only fair for this proposal. This is fair for whatever caching approach (primitive or pds) because it requires a key. This means that the cache has to be attached to something. Currently it's attached to operation descriptor, attributes, hint pd, device, number of threads, implementation ID, etc. If we want to change any of those we have to adjust the cache.